真偽値の字句解析が `true,` の時に `true` とならない

## 概要

`true,` や `true[` など空白や定義された記号が連続して続く場合に正くtokenを作れない。

## テストコード

```go
func TestBool(t *testing.T) {
	str := "true,"
	want := []parser.Tokener{
		parser.NewValueToken(parser.Bool, true),
		parser.NewMarkToken(parser.Comma),
	}
	got, _ := parser.Analyze([]byte(str))
	if !reflect.DeepEqual(got, want) {
		t.Errorf("str: %q, got: %v, want: %v", str, got, want)
	}

	str2 := "true["
	want2 := []parser.Tokener{
		parser.NewValueToken(parser.Bool, true),
		parser.NewMarkToken(parser.LeftSquareBracket),
	}
	got2, _ := parser.Analyze([]byte(str2))
	if !reflect.DeepEqual(got2, want2) {
		t.Errorf("str: %q, got: %v, want: %v", str2, got2, want2)
	}
}
```

## 実行結果

```shell
mini-parser-go on git main [$!] via go v1.22.5
➜ go test ./...
--- FAIL: TestBool (0.00s)
    analyzer_test.go:26: str: "true,", got: [{4 false}], want: [{4 true} {10}]
    analyzer_test.go:36: str: "true[", got: [{4 false}], want: [{4 true} {5}]
FAIL
FAIL    github.com/YO-RO/mini-parser-go/parser  0.308s
FAIL
```

--- a/parser/analyzer.go
+++ b/parser/analyzer.go
@@ -82,13 +82,13 @@ func extractStringAsToken(str string, startIdx int) (ValueToken, int, error) {
 	// 直前に\がない"
 	// または 直前に\が偶数回連続している"
 	// `"`や`\\"`などがマッチ
-	re := regexp.MustCompile(`(?:^|[^\\]|((^|[^\\])(\\\\)+))(")`)
+	re := regexp.MustCompile(`(?:^|[^\\]|(?:(?:^|[^\\])(?:\\\\)+))(")`)
 	loc := re.FindStringSubmatchIndex(str[startIdx+1:])
 	if loc == nil {
 		return ValueToken{}, 0, ErrSyntax
 	}
 	// idxsはstr[firstQuotationIdx+1]からのインデックスであるためfirstQuotationIdx+1を足す
-	endIdx := startIdx + 1 + loc[1]
+	endIdx := startIdx + 1 + loc[3]
 
 	value, err := strconv.Unquote(str[startIdx:endIdx])
 	if err != nil {

--- a/parser/analyzer.go
+++ b/parser/analyzer.go
@@ -152,14 +152,13 @@ func mayBeBool(str string, i int) bool {
 }
 
 func extractBoolAsToken(str string, startIdx int) (ValueToken, int, error) {
-	// ?i: はcase insentive
 	// ?: はグループをキャプチャしない
-	re := regexp.MustCompile(`(?i:true|false)(?:[\s,:"{}\[\]]|$)`)
+	re := regexp.MustCompile(`(true|false)(?:[\s,:"{}\[\]]|$)`)
 	loc := re.FindStringSubmatchIndex(str[startIdx:])
 	if loc == nil {
 		return ValueToken{}, 0, ErrUndefinedSymbol
 	}
-	endIdx := startIdx + loc[1]
+	endIdx := startIdx + loc[3]
 
 	var value bool
 	if str[startIdx:endIdx] == "true" {

--- a/parser/analyzer_test.go
+++ b/parser/analyzer_test.go
@@ -66,6 +66,15 @@ func TestBoolAnalyzer(t *testing.T) {
 			[]parser.Tokener{boolToken(t, false)},
 		},
 		{
+			"真(連続して記号が来る場合)",
+			"true,",
+			nil,
+			[]parser.Tokener{
+				boolToken(t, true),
+				parser.NewMarkToken(parser.Comma),
+			},
+		},
+		{
 			"定義されていないキーワード[trueeeee]",
 			"trueeeee",
 			parser.ErrUndefinedSymbol,

